### DISCUSSION TITLE
Unhandled promise rejection with TypeError message

### DISCUSSION THREAD

**Discussion by mvdgeijn** [reactphp/reactphp#530] [Q&A]
Hi,

I'm using this repository through the beyondcode laravel-websockets library, but I'm not sure where to report a problem I'm running into. I get the following error trace:

Unhandled promise rejection with TypeError: React\Http\Io\ClientRequestStream::closeError(): Argument #1 ($error) must be of type Exception, Error given, called in C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\promise\src\Internal\RejectedPromise.php on line 73 in C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\ClientRequestStream.php:238
Stack trace:
#0 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\promise\src\Internal\RejectedPromise.php(73): React\Http\Io\ClientRequestStream->closeError(Object(Error))
#1 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\promise\src\Promise.php(43): React\Promise\Internal\RejectedPromise->then(Object(Closure), Array)
#2 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\ClientRequestStream.php(100): React\Promise\Promise->then(Object(Closure), Array)
#3 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\ClientRequestStream.php(122): React\Http\Io\ClientRequestStream->writeHead()
#4 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\ClientRequestStream.php(135): React\Http\Io\ClientRequestStream->write('{"app_id":"15...')
#5 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\Sender.php(152): React\Http\Io\ClientRequestStream->end('{"app_id":"15...')
#6 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\Transaction.php(146): React\Http\Io\Sender->send(Object(React\Http\Message\Request))
#7 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Io\Transaction.php(83): React\Http\Io\Transaction->next(Object(React\Http\Message\Request), Object(React\Promise\Deferred), Object(React\Http\Io\ClientRequestState))
#8 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Browser.php(854): React\Http\Io\Transaction->send(Object(React\Http\Message\Request))
#9 C:\Users\Marc\PhpstormProjects\mainsite\vendor\react\http\src\Browser.php(176): React\Http\Browser->requestMayBeStreaming('POST', 'http://www.b...', Array, Object(GuzzleHttp\Psr7\Stream))

...

The trace continues, but seems to be irrelevant. Is this an issue to report here? Or should it be reported to the calling library?

Kind regards
Marc

**Comment by WyriHaximus** [comment]
This is a fun one so to say. Essentially when https://github.com/orgs/reactphp/discussions/481 is in place you'll get the underlying `Error` instead of the `TypeError` you are seeing now. However, that error is coming from somewhere and has to be addressed. If this is easily reproducible I'd suggest to temporary remove the type hint and see what the underlying error is.

**Comment by SimonFrings** [comment]
@mvdgeijn Thanks for bringing this up :+1:

> Is this an issue to report here?

This topic has to do with our [Promise v3](https://github.com/reactphp/promise), so I think it's valid to bring this to our attention. With the [release of Promise v3](https://github.com/reactphp/promise/releases/tag/v3.0.0) we now [report any unhandled promise rejection](https://github.com/reactphp/promise/pull/248) instead of suppressing this information. We're currently not aware of anything broken with this that would lead to this type of message, but if you can recreate this behavior with a reproducible example using our components, then we'll have a closer look.

> Or should it be reported to the calling library?

This could be a good idea in my opinion. From the message above, it seems that the beyondcode laravel-websockets library is currently not fully compatible with Promise v3, as there are unhandled promise rejections being reported. As a transitional solution you could use a version that installs an older Promise version until this is fixed. This aligns with what @WyriHaximus mentioned about examining the underlying error to identify the cause of the problem.

I hope this helps :)
